// ngx-bootstrap-navbar/esm2015/lib/navbar-dynamic-collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ViewportRuler } from '@angular/cdk/scrolling';
import { Directive, ElementRef, NgZone, ChangeDetectorRef, } from '@angular/core';
import { merge, Subject } from 'rxjs';
import { distinctUntilChanged, map, takeUntil, filter } from 'rxjs/operators';
export class NgxNavbarDynamicExpandDirective {
    /**
     * @param {?} _viewportRuler
     * @param {?} _elRef
     * @param {?} _ngZone
     * @param {?} _cdRef
     */
    constructor(_viewportRuler, _elRef, _ngZone, _cdRef) {
        this._viewportRuler = _viewportRuler;
        this._elRef = _elRef;
        this._ngZone = _ngZone;
        this._cdRef = _cdRef;
        this._isExpanded = false;
        this._onDestroy = new Subject();
        this._update = new Subject();
        this.isExpanded$ = merge(this._viewportRuler.change(150), this._update).pipe(filter((/**
         * @return {?}
         */
        () => checkBootstrapStylesAreLoaded(this._elRef.nativeElement))), map((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const element = this._elRef.nativeElement;
            /** @type {?} */
            let overflowSize;
            if (this.isExpanded) {
                overflowSize = element.scrollWidth - element.offsetWidth;
            }
            else {
                /** @type {?} */
                const clone = (/** @type {?} */ (element.cloneNode(true)));
                clone.classList.add('navbar-expand');
                element.parentElement.appendChild(clone);
                overflowSize = clone.scrollWidth - clone.offsetWidth;
                element.parentElement.removeChild(clone);
            }
            if (!overflowSize) {
                return true;
            }
        })), distinctUntilChanged(), takeUntil(this._onDestroy));
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.isExpanded$.subscribe((/**
             * @param {?} isExpanded
             * @return {?}
             */
            (isExpanded) => {
                this._ngZone.run((/**
                 * @return {?}
                 */
                () => {
                    this._isExpanded = isExpanded;
                    this._cdRef.markForCheck();
                }));
            }));
        }));
    }
    /**
     * @return {?}
     */
    get isExpanded() {
        return this._isExpanded;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            if (checkBootstrapStylesAreLoaded(this._elRef.nativeElement)) {
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    this.updateExpansion();
                }), 0);
            }
            else {
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    this.ngAfterContentInit();
                }), 1000 / 60);
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._update.complete();
        this._onDestroy.next();
        this._onDestroy.complete();
    }
    /**
     * @return {?}
     */
    updateExpansion() {
        this._update.next();
    }
}
NgxNavbarDynamicExpandDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxNavbarDynamicExpand]',
                host: {
                    class: 'navbar text-nowrap',
                    '[class.navbar-expand]': 'isExpanded',
                },
            },] }
];
/** @nocollapse */
NgxNavbarDynamicExpandDirective.ctorParameters = () => [
    { type: ViewportRuler },
    { type: ElementRef },
    { type: NgZone },
    { type: ChangeDetectorRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._isExpanded;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._onDestroy;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._update;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype.isExpanded$;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._viewportRuler;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._elRef;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._ngZone;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._cdRef;
}
/**
 * @param {?} element
 * @return {?}
 */
function checkBootstrapStylesAreLoaded(element) {
    /** @type {?} */
    const computedStyle = getComputedStyle(element);
    return computedStyle && computedStyle.whiteSpace === 'nowrap';
}
