// ngx-bootstrap-navbar/esm2015/lib/navbar-collapse.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ViewEncapsulation } from '@angular/core';
import { ngxNavbarAnimations } from './navbar-animation';
export class NgxNavbarCollapseComponent {
    /**
     * @param {?} _cdr
     */
    constructor(_cdr) {
        this._cdr = _cdr;
        this.isCollapsed = true;
        this.isTransitioning = false;
    }
    /**
     * @return {?}
     */
    toggle() {
        this.isTransitioning = true;
        this.isCollapsed = !this.isCollapsed;
        this._cdr.markForCheck();
    }
    /**
     * @return {?}
     */
    close() {
        if (this.isCollapsed) {
            return;
        }
        this.isTransitioning = true;
        this.isCollapsed = true;
        this._cdr.markForCheck();
    }
    /**
     * @return {?}
     */
    open() {
        if (!this.isCollapsed) {
            return;
        }
        this.isTransitioning = true;
        this.isCollapsed = false;
        this._cdr.markForCheck();
    }
}
NgxNavbarCollapseComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-navbar-collapse',
                template: `
    <ng-content></ng-content>
  `,
                host: {
                    class: 'navbar-collapse ngx-navbar-collapse',
                    '[class.collapse]': '!isTransitioning',
                    '[class.show]': '!isCollapsed && !isTransitioning',
                    '[class.collapsing]': 'isTransitioning',
                    '[@slideState]': '!isCollapsed',
                    '(@slideState.done)': 'isTransitioning = false'
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                animations: [ngxNavbarAnimations.navbarSlide],
                exportAs: 'ngxNavbarCollapse',
                styles: [`
      .ngx-navbar-collapse.collapsing {
        height: auto;
      }
    `]
            }] }
];
/** @nocollapse */
NgxNavbarCollapseComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
if (false) {
    /** @type {?} */
    NgxNavbarCollapseComponent.prototype.isCollapsed;
    /** @type {?} */
    NgxNavbarCollapseComponent.prototype.isTransitioning;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarCollapseComponent.prototype._cdr;
}
