// ngx-bootstrap-navbar/esm5/lib/navbar-collapse.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ViewEncapsulation } from '@angular/core';
import { ngxNavbarAnimations } from './navbar-animation';
var NgxNavbarCollapseComponent = /** @class */ (function () {
    function NgxNavbarCollapseComponent(_cdr) {
        this._cdr = _cdr;
        this.isCollapsed = true;
        this.isTransitioning = false;
    }
    /**
     * @return {?}
     */
    NgxNavbarCollapseComponent.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this.isTransitioning = true;
        this.isCollapsed = !this.isCollapsed;
        this._cdr.markForCheck();
    };
    /**
     * @return {?}
     */
    NgxNavbarCollapseComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        if (this.isCollapsed) {
            return;
        }
        this.isTransitioning = true;
        this.isCollapsed = true;
        this._cdr.markForCheck();
    };
    /**
     * @return {?}
     */
    NgxNavbarCollapseComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        if (!this.isCollapsed) {
            return;
        }
        this.isTransitioning = true;
        this.isCollapsed = false;
        this._cdr.markForCheck();
    };
    NgxNavbarCollapseComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-navbar-collapse',
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        class: 'navbar-collapse ngx-navbar-collapse',
                        '[class.collapse]': '!isTransitioning',
                        '[class.show]': '!isCollapsed && !isTransitioning',
                        '[class.collapsing]': 'isTransitioning',
                        '[@slideState]': '!isCollapsed',
                        '(@slideState.done)': 'isTransitioning = false'
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [ngxNavbarAnimations.navbarSlide],
                    exportAs: 'ngxNavbarCollapse',
                    styles: ["\n      .ngx-navbar-collapse.collapsing {\n        height: auto;\n      }\n    "]
                }] }
    ];
    /** @nocollapse */
    NgxNavbarCollapseComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return NgxNavbarCollapseComponent;
}());
export { NgxNavbarCollapseComponent };
if (false) {
    /** @type {?} */
    NgxNavbarCollapseComponent.prototype.isCollapsed;
    /** @type {?} */
    NgxNavbarCollapseComponent.prototype.isTransitioning;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarCollapseComponent.prototype._cdr;
}
