// ngx-bootstrap-navbar/esm5/lib/navbar-dynamic-collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ViewportRuler } from '@angular/cdk/scrolling';
import { Directive, ElementRef, NgZone, ChangeDetectorRef, } from '@angular/core';
import { merge, Subject } from 'rxjs';
import { distinctUntilChanged, map, takeUntil, filter } from 'rxjs/operators';
var NgxNavbarDynamicExpandDirective = /** @class */ (function () {
    function NgxNavbarDynamicExpandDirective(_viewportRuler, _elRef, _ngZone, _cdRef) {
        var _this = this;
        this._viewportRuler = _viewportRuler;
        this._elRef = _elRef;
        this._ngZone = _ngZone;
        this._cdRef = _cdRef;
        this._isExpanded = false;
        this._onDestroy = new Subject();
        this._update = new Subject();
        this.isExpanded$ = merge(this._viewportRuler.change(150), this._update).pipe(filter((/**
         * @return {?}
         */
        function () { return checkBootstrapStylesAreLoaded(_this._elRef.nativeElement); })), map((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var element = _this._elRef.nativeElement;
            /** @type {?} */
            var overflowSize;
            if (_this.isExpanded) {
                overflowSize = element.scrollWidth - element.offsetWidth;
            }
            else {
                /** @type {?} */
                var clone = (/** @type {?} */ (element.cloneNode(true)));
                clone.classList.add('navbar-expand');
                element.parentElement.appendChild(clone);
                overflowSize = clone.scrollWidth - clone.offsetWidth;
                element.parentElement.removeChild(clone);
            }
            if (!overflowSize) {
                return true;
            }
        })), distinctUntilChanged(), takeUntil(this._onDestroy));
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.isExpanded$.subscribe((/**
             * @param {?} isExpanded
             * @return {?}
             */
            function (isExpanded) {
                _this._ngZone.run((/**
                 * @return {?}
                 */
                function () {
                    _this._isExpanded = isExpanded;
                    _this._cdRef.markForCheck();
                }));
            }));
        }));
    }
    Object.defineProperty(NgxNavbarDynamicExpandDirective.prototype, "isExpanded", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isExpanded;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxNavbarDynamicExpandDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            if (checkBootstrapStylesAreLoaded(_this._elRef.nativeElement)) {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.updateExpansion();
                }), 0);
            }
            else {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.ngAfterContentInit();
                }), 1000 / 60);
            }
        }));
    };
    /**
     * @return {?}
     */
    NgxNavbarDynamicExpandDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._update.complete();
        this._onDestroy.next();
        this._onDestroy.complete();
    };
    /**
     * @return {?}
     */
    NgxNavbarDynamicExpandDirective.prototype.updateExpansion = /**
     * @return {?}
     */
    function () {
        this._update.next();
    };
    NgxNavbarDynamicExpandDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxNavbarDynamicExpand]',
                    host: {
                        class: 'navbar text-nowrap',
                        '[class.navbar-expand]': 'isExpanded',
                    },
                },] }
    ];
    /** @nocollapse */
    NgxNavbarDynamicExpandDirective.ctorParameters = function () { return [
        { type: ViewportRuler },
        { type: ElementRef },
        { type: NgZone },
        { type: ChangeDetectorRef }
    ]; };
    return NgxNavbarDynamicExpandDirective;
}());
export { NgxNavbarDynamicExpandDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._isExpanded;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._onDestroy;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._update;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype.isExpanded$;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._viewportRuler;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._elRef;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._ngZone;
    /**
     * @type {?}
     * @private
     */
    NgxNavbarDynamicExpandDirective.prototype._cdRef;
}
/**
 * @param {?} element
 * @return {?}
 */
function checkBootstrapStylesAreLoaded(element) {
    /** @type {?} */
    var computedStyle = getComputedStyle(element);
    return computedStyle && computedStyle.whiteSpace === 'nowrap';
}
